packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts

```diff
@@ -1108,4 +1124,79 @@ See documentation: https://www.algolia.com/doc/guides/building-search-ui/going-f
 
     global.document = originalDocument;
   });
+
+  test('does not immediately set userToken if a rerender is expected', async () => {
+    const searchClient = createSearchClient({
+      search: jest.fn((requests) => {
+        return Promise.resolve(
+          createMultiSearchResponse<any>(
+            ...requests.map(() =>
+              createSingleSearchResponse({
+                renderingContent: {
+                  facetOrdering: {
+                    facets: {
+                      order: ['brand'],
+                    },
+                  },
+                },
+                hits: [{ objectID: '1' }],
+              })
+            )
+          )
+        );
+      }),
+    }) as SearchClient & { search: jest.Mock };
+
+    const { insightsClient, instantSearchInstance, getUserToken } =
+      createTestEnvironment({ searchClient, started: false });
+
+    insightsClient('init', { partial: true, anonymousUserToken: true });
+    instantSearchInstance.use(createInsightsMiddleware({ insightsClient }));
+
+    instantSearchInstance.addWidgets([
+      dynamicWidgets({
+        container: document.createElement('div'),
+        widgets: [
+          (container) => refinementList({ container, attribute: 'brand' }),
+        ],
+      }),
+      hits({
+        container: document.createElement('div'),
+      }),
+    ]);
+
+    instantSearchInstance.start();
+
+    await wait(0);
+
+    // Dynamic widgets will trigger 2 searches. To avoid missing the cache on the second search, createInsightsMiddleware delays setting the userToken.
+
+    expect(searchClient.search).toHaveBeenCalledTimes(2);
+    expect(
+      searchClient.search.mock.calls.map(([[call]]) => call.params.userToken)
+    ).toMatchInlineSnapshot(`
+      [
+        undefined,
+        undefined,
+      ]
+    `);
```

Comment:
maybe this should be `searchClient.search.mock.calls[0][0].params.userToken).toBeUndefined()` and (`[1]`)? it's somewhat clear, but possibly a bit much logic for a test

Reply:
This is clearer for me as well. I've updated the test.